### PR TITLE
Bug/487 digits groups separator ambiguity

### LINKED ISSUE #399: Can't Use Thousand Separators Without Using A Decimal Separator
Thousand separators in an amount number can't be used unless the number also has a decimal separator. If there is only one punctuation mark, it's always considered to be a decimal separator even if one or more of the following are present:
- A "D directive" with:
  - A sample amount using only a thousand separator.
  - A sample amount using both thousand and decimal separators.
- A commodity directive that defines a format similar to the sample amounts listed above.
- A preceding transaction with an amount that uses a format similar to the sample amounts listed above.

For example:

```
D JPY 111,111,111
# also doesn't work:
# D JPY 111,111,111.0
# D JPY 1,111
# D JPY 1,111.0

2016/01/01 default_commodity
  (asset:no_comma)  100
  (asset:comma)  100,500
```

The balance sheet for this journal adds up to ""200,500" or "200.500" for either commodity, rather than "100,600".

Using a commodity directive also has incorrect totals. An unexpected value rounding also occurs with the format "JPY 111,111,111".

```
commodity JPY
  format JPY 111,111,111 # also causes unexpected rounding

#  format JPY 111,111,111.0
#  format JPY 1,111
#  format JPY 1,111.0

2016/01/01 yen
  (asset:commodity:no_comma)  JPY 100
  (asset:commodity:comma)  JPY 100,500
```

Using a preceding dummy transaction with a sample amount has the same output as the cases above:

```
2016/01/01 init
  (asset:commodity:init)  JPY 111,111,111
  (asset:commodity:init)  JPY -111,111,111
  # same output with the formats tried in other examples

2016/01/01 yen
  (asset:commodity:no_comma)  JPY 100
  (asset:commodity:comma)  JPY 100,500
```
## 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/37935078-can-t-use-thousand-separators-without-using-a-decimal-separator?utm_campaign=plugin&utm_content=tracker%2F536505&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F536505&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ony**
Trust latest `commodity` or `D` (default commodity) directives in case of ambiguity in parsing number.

This kinda resolves #399 and its duplicate #487. We still may get issues caused by misinterpreting numbers `1,000` and `1.0` both as `1` if no `commodity` directive or default commodity `D` directive were provided in same journal file earlier.

### CODE REVIEW

File: hledger-lib/Hledger/Read/Common.hs [hunk 1]
```diff
@@ -145,6 +146,12 @@ setDefaultCommodityAndStyle cs = modify' (\j -> j{jparsedefaultcommodity=Just cs
 getDefaultCommodityAndStyle :: JournalParser m (Maybe (CommoditySymbol,AmountStyle))
 getDefaultCommodityAndStyle = jparsedefaultcommodity `fmap` get
 
+getDefaultDecimalHint :: JournalParser m (Maybe Char)
```
Comment (simonmichael) [inline hledger-lib/Hledger/Read/Common.hs]: Let's add short haddocks explaining what these do.

**Comment by simonmichael** [comment]
I trust your judgement that this works better. We should have haddocks for the two get*Style functions, and an update to the manual explaining how things work now.

**Comment by ony** [comment]
@simonmichael , as you probably noticed expecting that `format` directive will affect interpretation is natural. But actual behavior is pretty hard to document. Because mostly it something like:
"If you have issue with misinterpreting decimals try putting `commodity` directive in the begining of the file" and that was a first thought of me, and as I remember yours also.

If you think we should have a separate directive for that (probably still currency-specific) it should be easy to add that.

**Comment by simonmichael** [comment]
The doc looks reasonable to me. Merging, thank you!

**Pull request merged**